core/src/impl/Kokkos_Tuning.cpp

```diff
@@ -0,0 +1,167 @@
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#include <unordered_map>
+#include <unordered_set>
+#include <impl/Kokkos_Tuning.hpp>
+namespace Kokkos {
+namespace Tools {
+
+static std::unordered_map<size_t, std::unordered_set<size_t>>
+    features_per_context;
+static std::unordered_set<size_t> active_features;
+static std::unordered_map<size_t, VariableValue> feature_values;
+
+void declareTuningVariable(const std::string& variableName, size_t uniqID,
+                           VariableInfo info) {
+#ifdef KOKKOS_ENABLE_TUNING
+  if (tuningVariableDeclarationCallback != nullptr) {
+    (*tuningVariableDeclarationCallback)(variableName.c_str(), uniqID, info);
+  }
+#endif
+}
+
+void declareContextVariable(const std::string& variableName, size_t uniqID,
+                            VariableInfo info,
+                            Kokkos::Tools::SetOrRange candidate_values) {
+#ifdef KOKKOS_ENABLE_TUNING
+  if (contextVariableDeclarationCallback != nullptr) {
+    (*contextVariableDeclarationCallback)(variableName.c_str(), uniqID, info,
+                                          candidate_values);
+  }
+#endif
+}
+
+void declareContextVariableValues(size_t contextId, size_t count,
+                                  VariableValue* values) {
+#ifdef KOKKOS_ENABLE_TUNING
+  if (features_per_context.find(contextId) == features_per_context.end()) {
+    features_per_context[contextId] = std::unordered_set<size_t>();
+  }
```

Comment:
I believe this check is unnecessary since I think line 83 will default initialize the set and then insert into it.